### ISSUE TITLE
[Enhancement] Providing a way to specify a rule without the select box (i.e. Calling specific rules from extensions, keybindings etc.)

### ISSUE THREAD

**Issue by frnco** [bhughes339/vscode-replacerules#9]
Currently the only way to run a task is through the action `extension.chooseRule` that provides a selection list (a.k.a. Quick Pick), so a user can't create keybindings or automate running a specific rule.

Besides being the first feature listed in the RegReplace Docs, it could help a lot to automate and chain replacements (i.e. CSV header and formatting data in the same file. In my case I get files from Gimp and I wanna automate my checklist of replaces).

Exporting per-rule actions could be optional for backward compatibility, and after looking at the code I think it fits pretty nicely, so I'm considering making this, should I do a PR? If yes, what should I keep in mind? (Tests, gotchas etc.)

For now I am thinking something like this:
```
"replacerules.rules": [
    {
        "name": "Remove trailing and leading whitespace",
        "find": "^\\s*(.*)\\s*$",
        "replace": "$1",
        "ruleKey": "autoStrip"
    }
]
```
To do stuff like this:
`{"key": "ctrl+alt+s", "command": "extension.runRule", "args": {externalAction: "autoStip"}}`

Most code can be shared by refactoring `chooseRule` in 3 functions:
- Refresh rules and return the list of rules plus an object using `[ruleKey]: rule`, I would not create the `new ReplaceRule` yet;
- Select the rule either from the object if `ruleKey` was specified or from the list using QuickPick;
- Run `doReplace`.

The object could also have the keys pointing to an index in the array or whatever. To put in:

```
let rules = getRulesList();
let rule = getRule(rules, args: nil);
let ruleClass = new ReplaceRule(rule); // Maybe a function returning the ruleClass from a rule?
doReplace(ruleClass);
```

Could also be `doReplace(getRuleClass(getRule(getRulesList(), args: nil)));`, but this looks awful...

I can see a lot of benefits in refactoring for changing and adding features (RegReplace was listed as inspiration and has a lot of features so I guess it fits) i.e. having a function that returns the list of rules makes it trivial to search rules, backup, export or whatever, same for getting a rule, I can imagine dynamic rules that receive an argument to use in the RegEx and people adding employees' names that weren't in a CSV using emails or other info etc.

**Comment by bhughes339** [comment]
Thank you for the write-up! This was a feature I wanted to include from the get-go, but the main issue was finding a way to keep each key unique in a user-friendly way. I explored a few different methods of making each rule unique--mainly, requiring each rule object in the settings be associated with a unique key (`"autoStrip": {"name": "autostrip", [...]}`)--but ultimately dropped the idea because I didn't really have the time for it.

If you'd like to submit a refactor PR, that would be great. There aren't any real "gotchas" in the code, it's pretty basic (and I'm open to different coding styles and ideas...still learning Typescript 😄). Otherwise I'll get to this some time in the future. Thanks again!

**Comment by frnco** [comment]
@bhughes339 Yeah, having a key for each rule would pretty much solve it, but there should be a way to do this withou breaking stuff.

Another option is to generate IDs and use that, maybe using a hash of the rule or something..?

**Comment by bhughes339** [comment]
@frnco The rules really should be unique by name--~~and I believe there is a JSON schema option to make them unique~~ (edit: this is not possible)--but that would mean keybindings would be broken when changing the name of a rule. Generating ids and mapping them to the rules array is also an option, but vscode extensions are coded lazily, and it seems likely that the mapping would be garbage collected when the extension has been inactive for a bit.

**Comment by bhughes339** [comment]
I think I'm going to follow the Snippet standard and go with the key route, It's probably going to be obnoxious to convert existing rules to the new format, but I've wanted to add keybinding support for a while and this seems like the best way to do it.

IDEA: Create a "Convert to new format" function that will take the `name` field of each rule and move it to the key. Example:

Old format:
```json
{
    "replacerules.rules": [
        {
            "name": "LaTeX Quotation Convert",
            "find": "\"([^\"]*)\"",
            "replace": "``$1''",
            "languages": "latex"
        },
        {
            "name": "Remove leading whitespace",
            "find": "^\\s*(.*)\\s*$",
            "replace": "$1"
        }
    ]
}
```

New format:
```json
{
    "replacerules.rules": {
        "LaTeX Quotation Convert": {
            "find": "\"([^\"]*)\"",
            "replace": "``$1''",
            "languages": "latex"
        },
        "Remove leading whitespace": {
            "find": "^\\s*(.*)\\s*$",
            "replace": "$1"
        }
    }
}
```

**Comment by bhughes339** [comment]
This should be resolved in version 0.2.0. Let me know if there are any issues with the new functionality. Specifics can be found in the README and changelog. Thank you!
